### DISCUSSION TITLE
Accessing UIFactory.SeparatorTemplate and UIFactory.HeadAddition

### DISCUSSION THREAD

**Discussion by SeanAWalsh** [NetSparkleUpdater/NetSparkle#181] [Ideas]
Hi there.

I see in NetSparkle.UI.Winforms.NetFramework the UpdateAvailableWindow constructor has a "Head Addition" to allow one to splice a piece of HTML into the header of the release notes HTML. But it's never called and there is no way to use it. I'd like to use Markdown for my release notes and I'd like to have the following HTML in the head so that I don't have ye olde Times New Roman font:

"<style> body { font-family: Arial, Helvetica, sans-serif; } </style>".

But, of course you can't include HTML in Markdown (that's the point of Markdown, I guess). So I decided to use the Head Addition to inject this piece of HTML. While I was at it, I added SeparatorTemplate even though I have no need to replace it.

public class UIFactory : IUIFactory
{
	...
	
	/// <inheritdoc/>
	public bool HideRemindMeLaterButton { get; set; }

	**/// <inheritdoc/>
	public string SeparatorTemplate { get; set; }

	/// <inheritdoc/>
	public string HeadAddition { get; set; }**

	/// <inheritdoc/>
	public virtual IUpdateAvailable CreateUpdateAvailableWindow(SparkleUpdater sparkle, List<AppCastItem> updates, bool isUpdateAlreadyDownloaded = false)
	{
		var window = new UpdateAvailableWindow(sparkle, updates, _applicationIcon, isUpdateAlreadyDownloaded, **SeparatorTemplate, HeadAddition**);
		...
	}
	
	...
}

which means I can now do this in my code:

private void NetSparkleCheckForUpdates()
{
    _sparkle = new SparkleUpdater("myurl/appcast.xml", new Ed25519Checker(...))
    {
        UIFactory = new NetSparkleUpdater.UI.WinForms.UIFactory(Icon)
        {
            HeadAddition = "<style> body { font-family: Arial, Helvetica, sans-serif; } </style>"
        },
        ShowsUIOnMainThread = true
    };

    _sparkle.StartLoop(true, true);
}

Of course, I now have a non-standard NetSparkle. Is this the right way to handle this? Have I missed something? If this is, indeed a good idea, how does one go about getting it added to NetSparkle so I can get back to your wonderful standard? I suppose it should also be added to NetSparkleUpdater.UI.WinForms.NetCore code?

Cheers

Sean

While we're at it, I also did this...

        /// <inheritdoc/>
        **private** UIFactory()
        {
            HideReleaseNotes = false;
            HideRemindMeLaterButton = false;
            HideSkipButton = false;
            **SeparatorTemplate = string.Empty;
            HeadAddition = string.Empty;**
        }

        /// <inheritdoc/>
        public UIFactory(Icon applicationIcon) **: this()**
        {
            _applicationIcon = applicationIcon;
        }

A little bit of constructor chaining to save having to duplicate the initialisation code... (And Occam's razor suggested that I should mark the parameterless constructor private as it's never directly called)

**Comment by SeanAWalsh** [comment]
Just spotted the Issues section (new to GitHub) maybe ths belongs there...

**Reply by Deadpikle** [reply]
No worries. Welcome to GitHub! :)

I'll hopefully get to this in the next few days. Thanks for your message!

**Reply by SeanAWalsh** [reply]
Thanks for this. I can see you have your hands full. I'm happy to get involved a bit and I'm a reasonably sharp and tidy programmer but I guess you guys have to learn to trust me. Not sure how this all collaboration works without anarchy. I have another suggestion (and a fix). Should I raise it as a discussion or an issue?

**Reply by Deadpikle** [reply]
Generally speaking, on GitHub, you can file an issue for a suggestion and/or proposed fix. The discussion section is more for general questions ("how do I do XYZ?") or chit chat. If it's more things like what you posted to this discussion, go ahead an open an Issue instead, please.

Depending on how certain you are of a given fix, you can also open a pull request (also called a merge request) that references your issue (by number) without feedback from a person on the repository. Each issue/pull request has a given number and you can reference these on GitHub with a pound sign (#) followed by the number. For instance, this is discussion #181. A pull request lets you suggest changes to actual files, and they can be reviewed and merged in more easily along with giving the change author the credit. I've been needing to get around to handling #171 for a few weeks now (and it's bad of me for procrastinating this long on it).

(If you knew all that already, apologies. If you're not familiar with git or github at all, I suggest following this tutorial: https://product.hubspot.com/blog/git-and-github-tutorial-for-beginners)

> Not sure how this all collaboration works without anarchy.

Yeah...I am the only person right now who can merge in actual code; sort of the [BDFL](https://en.wikipedia.org/wiki/Benevolent_dictator_for_life) until someone else comes along and joins in 😆. On larger repositories, e.g. https://github.com/AvaloniaUI/Avalonia, you'd have multiple people in control, merging in code, handling issues, merging in pull requests, etc.

**Reply by SeanAWalsh** [reply]
Wow. Thanks for taking so much time to explain. I use Subversion a lot. But never Git and it's a bit different for sure. Seems to be a sort of two-hop thing going on with, as you say, a review and acceptance phase. Seems sensible to me.

I'll watch the tutorial. But in the meantime, I've prepared a small "issue" and a suggested set of code changes to handle it. I'll post it now as an issue so you know what's coming down the pipeline. I'll then try to push it through to you as a pull request.

By the end of that, I'll have a better understanding of what's going on. On the list of open issues, is there any particular one you'd like me to have a look at that doesn't cut across too much else you're doing?

**Reply by SeanAWalsh** [reply]
In fact, maybe I'll try a pull request for the thing raised at the start of this discussion as it only involves an edit to one file. Then you can review and accept if you like it.

**Reply by Deadpikle** [reply]
> In fact, maybe I'll try a pull request for the thing raised at the start of this discussion as it only involves an edit to one file. Then you can review and accept if you like it.

Go ahead! I probably won't get to it for a day or three, but I will do my best to respond sooner rather than later.

**Reply by Deadpikle** [reply]
> On the list of open issues, is there any particular one you'd like me to have a look at that doesn't cut across too much else you're doing?

Thanks for asking. Would need to look, but have run out of time for the evening, so will get back to this later if you haven't found things yourself.

**Reply by SeanAWalsh** [reply]
> > In fact, maybe I'll try a pull request for the thing raised at the start of this discussion as it only involves an edit to one file. Then you can review and accept if you like it.
> 
> Go ahead! I probably won't get to it for a day or three, but I will do my best to respond sooner rather than later.

No rush!

**Comment by Deadpikle** [comment]
> Is this the right way to handle this? 

Technically, you can create your own UIFactory class instead, rather than overwriting the built-in one.

```csharp
public class MyUIFactory : NetSparkleUpdater.UI.WinForms.UIFactory
{
    public string SeparatorTemplate { get; set; }

    public string HeadAddition { get; set; }

    // your constructor here if necessary...

    public override virtual IUpdateAvailable CreateUpdateAvailableWindow(SparkleUpdater sparkle, List<AppCastItem> updates, bool isUpdateAlreadyDownloaded = false)
    {
        var window = new UpdateAvailableWindow(sparkle, updates, _applicationIcon, isUpdateAlreadyDownloaded);
        if (HideReleaseNotes)
        {
            (window as IUpdateAvailable).HideReleaseNotes();
        }
        if (HideSkipButton)
        {
            (window as IUpdateAvailable).HideSkipButton();
        }
        if (HideRemindMeLaterButton)
        {
            (window as IUpdateAvailable).HideRemindMeLaterButton();
        }
        return window;
    }
}
```

Then just pass in that UIFactory subclass to `NetSparkle`.

There's already default parameters for the properties that eventually get sent to the `ReleaseNotesGrabber`, which sets up all the HTML, as you've seen.

Clearly there needs to be a little API rework here to make the separator template and head addition properties more easily settable/modifiable across all the UI libs, which probably means pulling them into the `UIFactory`, as you've noted. The way to modify these things is technically possible right now but not very visible to the end-user.

**Reply by SeanAWalsh** [reply]
Gotcha. Of course. The meek shall inherit...

**Reply by SeanAWalsh** [reply]
Does mean some not very nice code duplication though... :(

**Comment by Deadpikle** [comment]
This is now working in `2.0.0-preview20210404001`!

**Reply by Deadpikle** [reply]
I submitted too early 😵 I also made some changes to `ReleaseNotesGrabber` and a few other things to make things easier to adjust in the future.

**Reply by SeanAWalsh** [reply]
Many thanks!

**Reply by SeanAWalsh** [reply]
Nice job, by the way.
